[OpenMP] Patch for Support to loop bind clause : Checking Parent Region

SunilKuravinakop opened the pull request:
Changes uploaded to the phabricator on Dec 16th are lost because the phabricator is down. Hence re-uploading it to the github.com.

  Changes to be committed:
 	modified:   clang/include/clang/Sema/Sema.h
 	modified:   clang/lib/Sema/SemaOpenMP.cpp
 	modified:   clang/test/OpenMP/generic_loop_ast_print.cpp
 	modified:   clang/test/OpenMP/loop_bind_messages.cpp
 	modified:   clang/test/PCH/pragma-loop.cpp

Commit:
Changes uploaded to the phabricator on Dec 16th are lost because the
phabricator is down. Hence re-uploading it to the github.com.

  Changes to be committed:
 	modified:   clang/include/clang/Sema/Sema.h
 	modified:   clang/lib/Sema/SemaOpenMP.cpp
 	modified:   clang/test/OpenMP/generic_loop_ast_print.cpp
 	modified:   clang/test/OpenMP/loop_bind_messages.cpp
 	modified:   clang/test/PCH/pragma-loop.cpp

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -5072,6 +5072,18 @@ static bool checkNestingOfRegions(Sema &SemaRef, const DSAStackTy *Stack,
+    // Checks needed for mapping "loop" construct. Please check mapLoopConstruct
+    // for a detailed explanation
+    if (SemaRef.LangOpts.OpenMP >= 50 && CurrentRegion == OMPD_loop &&
+        ((BindKind == OMPC_BIND_parallel) || (BindKind == OMPC_BIND_teams)) &&
```
Drop extra parens

alexey-bataev commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -11335,6 +11335,7 @@ class Sema final {
   /// on the parameter of the bind clause. In the methods for the
   /// mapped directives, check the parameters of the lastprivate clause.
   bool checkLastPrivateForMappedDirectives(ArrayRef<OMPClause *> Clauses);
+
```
Unrelated change, drop it, make it in a separate NFC patch

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -6213,26 +6226,35 @@ StmtResult Sema::ActOnOpenMPExecutableDirective(
+  // Variable used to note down the DirectiveKind because mapLoopConstruct may
+  // change "Kind" variable, due to mapping of "omp loop" to other directives.
+  OpenMPDirectiveKind DK = Kind;
+  if ((Kind == OMPD_loop) || (PrevMappedDirective == OMPD_loop)) {
```
Extra parens

Commit:
Removing extra paranthesis.
Removing an empty line which was added to seperate a function from
another. This was the only change in the file and hence will be adding
it as a separate NFC patch as suggested by Alexey.

 Changes to be committed:
	modified:   clang/include/clang/Sema/Sema.h
	modified:   clang/lib/Sema/SemaOpenMP.cpp

Commit:
Merge branch 'main' into parent_loop_bind

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -6124,35 +6136,36 @@ processImplicitMapsWithDefaultMappers(Sema &S, DSAStackTy *Stack,
-                 ParentDirective == OMPD_target_parallel) {
+                 ParentDirective == OMPD_target_parallel)
         BindKind = OMPC_BIND_parallel;
-      } else if (ParentDirective == OMPD_teams ||
-                 ParentDirective == OMPD_target_teams) {
-        BindKind = OMPC_BIND_teams;
-      }
```
Why did you removed this? Remind me if we discussed it before.

Commit:
In case of unspecified bind in "#pragma omp loop" and parent being "teams" or "target teams", I had modified the if condition to use "isOpenMPTeamsDirective()" for the checks. However, this proved too restrictive and hence removed the check. Now, on Alexey pointing out the error, re-introduced the check to use "teams" or "target teams".
In the test case loop_bind_messages.cpp, Combined multiple tests into fewer functions.

  Changes to be committed:
 	modified:   clang/lib/Sema/SemaOpenMP.cpp
 	modified:   clang/test/OpenMP/loop_bind_messages.cpp

Commit:
Merge branch 'main' into parent_loop_bind

Commit:
Merge branch 'main' into parent_loop_bind

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -6124,35 +6136,39 @@ processImplicitMapsWithDefaultMappers(Sema &S, DSAStackTy *Stack,
         BindKind = OMPC_BIND_parallel;
-      } else if (ParentDirective == OMPD_teams ||
-                 ParentDirective == OMPD_target_teams) {
-        BindKind = OMPC_BIND_teams;
-      }
+    } else if (ParentDirective == OMPD_teams ||
+               ParentDirective == OMPD_target_teams) {
+      BindKind = OMPC_BIND_teams;
```
Please restore original code here

Commit:
Correcting the brackets related to teams in SemaOpenMP.
Changes to testcase to reflect the changed code.

  Changes to be committed:
	modified:   clang/lib/Sema/SemaOpenMP.cpp
 	modified:   clang/test/OpenMP/generic_loop_ast_print.cpp

Commit:
Merge branch 'main' into parent_loop_bind

alexey-bataev reviewed (approved):
LG
